File: DStructures/heap.js
class Heap {
  constructor() {
    this.dataList = [];
    this.size = 0;
  }

  // Insert New Node Into The Heap
  insertNode(data) {
    let i = this.size;
    this.dataList[i] = data;
    this.size++;

    let parentIdx = Math.floor((i - 1) / 2);

    while (i != 0 && this.dataList[i] < this.dataList[parentIdx]) {
      this.dataList[i] = this.dataList[parentIdx];
      this.dataList[parentIdx] = data;
      i = parentIdx;

      parentIdx = Math.floor((i - 1) / 2);
    }
  }

  // Print The Heap Method
  printList() {
    let print_data = "";
    for (let i = 0; i < this.size; ++i) {
      print_data += this.dataList[i] + " - ";
    }
    console.log(print_data);
  }

  // Get The Size Of The Heap
  getSize() {
    return this.size;
  }
}

let heap = new Heap();

heap.insertNode(5);
heap.insertNode(3);
heap.insertNode(8);
heap.insertNode(1);
heap.insertNode(2);
heap.insertNode(10);

heap.printList();

File: DStructures/heap.py
class Heap:
    def __init__(self) -> None:
        self.dataList = []
        self.size = 0

    def insert_node(self, data):
        i = self.size
        self.dataList[i] = data
        self.size += 1

        parent_index = (i - 1) // 2

        while i != 0 and self.dataList[i] < self.dataList[parent_index]:
            self.dataList[i], self.dataList[parent_index] = self.dataList[parent_index], data 
            i = parent_index

            parent_index = (i - 1) // 2

    
    def get_size(self):
        return self.size
    
    def print_list(self):
        print_data = "" 
        for i in self.dataList:
            print_data += i + " - " 
        print(print_data)


heap = Heap()

# Test inserting nodes
heap.insert_node(5)
heap.insert_node(3)
heap.insert_node(8)
heap.insert_node(1)
heap.insert_node(10)

# Test get_size method
assert heap.get_size() == 5

# Test print_list method
heap.print_list()  